STORM-1167: Add windowing support for storm core

arunmahadevan opened the pull request:
Currently, topologies that needs windowing support requires writing custom logic inside bolts making it tedious to handle the windowing and acking logic with custom logic.

The PR proposes to add framework level support to core storm bolts to process tuples in a time or a count based window. Sliding and tumbling windows based on tuple count or time duration are supported.

A new bolt interface is added for bolts that needs windowing support.

```
public interface IWindowedBolt extends IComponent {
    void prepare(Map stormConf, TopologyContext context, OutputCollector collector);
    /**
     * Process tuples falling within the window and optionally emit 
     * new tuples based on the tuples in the input window.
     */
    void execute(TupleWindow inputWindow);
    void cleanup();
}
```

`TupleWindow` gives access to the current tuples in the window, the tuples that expired and the new tuples that are added since last window was computed which will be useful for efficient windowing computations.

The `TopologyBuilder` wraps the IWindowedBolt implementation in an internal bolt `WindowedBoltExecutor` (similar to `BasicBoltExecutor`), which receives the tuples and hands it off to IWindowedBolt when the window conditions are met. The tuples are automatically acked when they fall out of the window and the tuples in a window are automatically anchored with the tuple emitted from an IWindowedBolt to provide atleast once guarentee. An example topology `SlidingWindowTopology` shows how to use the apis to compute a sliding window sum.

The tuples are tracked with the system timestamp when they are received by the bolt. Later we can add support for tuple based timestamps (e.g the ts when the tuple was generated). There is also scope for minimizing duplicates by tracking the evaluated tuples (based on some id) in a persistent store.

Similiar windowing constructs needs to be added to storm trident apis as well. That can be addressed once the basic support to store core is added and the common logic can be reused.

revans2 commented on examples/storm-starter/src/jvm/storm/starter/SlidingWindowTopology.java:
```diff
@@ -0,0 +1,129 @@
+        @Override
+        public void nextTuple() {
+            Utils.sleep(100);
+            Random rand = new Random();
```
Minor nit, declaring a new Random each time is not really good practice.

arunmahadevan replied:
Yes, I overlooked this when I copied the code from another bolt. Will fix it.

revans2 commented on examples/storm-starter/src/jvm/storm/starter/SlidingWindowTopology.java:
```diff
@@ -0,0 +1,129 @@
+            for (Tuple tuple : inputWindow.getExpired()) {
+                sum -= (int) tuple.getValue(0);
+            }
+            collector.emit(new Values(sum));
```
If this is a sliding window, why are we keeping the sum around between windows?

Parth-Brahmbhatt replied:
I think this is trying to show an optimization where we dont calculate the entire sum each time, instead we just add the new values and subtract the expired one.

revans2 replied:
Oh I see you have the `getNew()` and the `getExpired()`.  OK It would be good to add a comment on that, because it was very confusing to me when I just glanced through the code.

arunmahadevan replied:
I will add more comments in the example topology explaining the details.

revans2 commented on examples/storm-starter/src/jvm/storm/starter/SlidingWindowTopology.java:
```diff
@@ -0,0 +1,129 @@
+        TopologyBuilder builder = new TopologyBuilder();
+        builder.setSpout("integer", new RandomIntegerSpout(), 1);
+        builder.setBolt("window", new SlidingWindowSumBolt().withWindow(new Count(30), new Count(10)),
+                        //new SlidingWindowSumBolt().withTumblingWindow(new Duration(20, TimeUnit.SECONDS))
```
If we want a separate example with a TumblingWindow we should probably have that instead of a comment in the code.

arunmahadevan replied:
Will add a separate Bolt in that shows the usage of TumblingWindow.

revans2 commented on storm-core/src/jvm/backtype/storm/topology/IWindowedBolt.java:
```diff
@@ -0,0 +1,40 @@
+public interface IWindowedBolt extends IComponent {
+    /**
+     * This is similar to the {@link backtype.storm.task.IBolt#prepare(Map, TopologyContext, OutputCollector)} except
+     * that while emitting the tuples are are automatically anchored to the tuples in the inputWindow.
```
```are are``` We probably need only one.

revans2 commented on storm-core/src/jvm/backtype/storm/topology/WindowedBoltExecutor.java:
```diff
@@ -0,0 +1,200 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p/>
```
minor nit. The `<p/>` is not in other license files, I think IntelleJ is adding it in for you.

arunmahadevan replied:
I will fix it across all files.

revans2 commented on storm-core/src/jvm/backtype/storm/topology/WindowedBoltExecutor.java:
```diff
@@ -0,0 +1,200 @@
+        }
+    }
+
+    // TODO: add more validation
```
If we need more validation we should add it in.  If we don't we should remove this comment.

Parth-Brahmbhatt commented on examples/storm-starter/src/jvm/storm/starter/SlidingWindowTopology.java:
```diff
@@ -0,0 +1,129 @@
+
+        @Override
+        public void execute(TupleWindow inputWindow) {
+            System.out.println("Events in current window: " + inputWindow.get().size());
```
remove sysout

revans2 commented on storm-core/src/jvm/backtype/storm/topology/WindowedBoltExecutor.java:
```diff
@@ -0,0 +1,200 @@
+        bolt.prepare(stormConf, context, windowedOutputCollector);
+        this.listener = newWindowLifecycleListener();
+        this.windowManager = initWindowManager(listener, stormConf);
+        LOG.info("Initialized window manager {} ", this.windowManager);
```
This looks more like a debug statement to me.

revans2 commented on storm-core/src/jvm/backtype/storm/topology/base/BaseWindowedBolt.java:
```diff
@@ -0,0 +1,184 @@
+public class BaseWindowedBolt implements IWindowedBolt {
+    private static final Logger LOG = LoggerFactory.getLogger(BaseWindowedBolt.class);
+
+    private transient Map<String, Object> windowConfiguration;
```
This this marked as transient, do the with... methods even work?  It will not be serialized and it looks like other configs override it anyways,

revans2 replied:
Never mind I see how you are doing it, and windowConfiguration is read by the topology builder.  Looks good.

Parth-Brahmbhatt replied:
I am not sure why we can't just have instance variable for all the configs? Why does it need to go to a transient map?

arunmahadevan replied:
I added new constants in Config and overriding it here via the componentConfiguration. This gives a way to define default values for these params at topology level if needed and override it at per bolt level. Defining it in same place with other params makes it more consistent and enables us to reuse the constants for trident as well. I think defining it in BaseWindowedBolt makes it more specific to core.

revans2 commented on storm-core/src/jvm/backtype/storm/windowing/CountEvictionPolicy.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class CountEvictionPolicy<T> implements EvictionPolicy<T> {
+    private final int threshold;
+    private AtomicInteger currentCount;
```
This should probably be a final too.

arunmahadevan replied:
Thanks, I missed it.

revans2 commented on storm-core/src/jvm/backtype/storm/topology/base/BaseWindowedBolt.java:
```diff
@@ -0,0 +1,184 @@
+
+    @Override
+    public void execute(TupleWindow inputWindow) {
+
```
Wouldn't it be better to have this abstract?

Parth-Brahmbhatt replied:
+1.

arunmahadevan replied:
Yes, makes sense to keep it abstract and not define the `execute` method, since that should be overridden by the implementor.

revans2 commented on storm-core/src/jvm/backtype/storm/topology/base/BaseWindowedBolt.java:
```diff
@@ -0,0 +1,184 @@
+
+    @Override
+    public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
+    }
```
Can you put in a comment in all the empty methods, something like ```\\NOOP``` just so it is obvious that it is intended to be left blank.

arunmahadevan replied:
Sure.

revans2 commented on storm-core/src/jvm/backtype/storm/windowing/CountTriggerPolicy.java:
```diff
@@ -0,0 +1,62 @@
+public class CountTriggerPolicy<T> implements TriggerPolicy<T> {
+    private final int count;
+    private AtomicInteger currentCount;
+    private TriggerHandler handler;
```
Should currentCount and handler be final as well?

arunmahadevan replied:
Thanks, I missed it will make it final.

revans2 commented:
On the surface this looks really good, a few minor nits with the code, but I have just done a quick pass through the code, and would like to dig in more before giving it a formal +1.

haohui commented:
The high level API looks good to me overall. It maintains a view of all the events in the window which is a powerful concept.

I have several questions on how this PR can help implement two common use cases.

(1) Aggregation (e.g., min / max) over a sliding window
(2) Stream joins over a large amount of data

The abstractions of views in memory are insufficient because for an efficient algorithm for (1) does not need every single events in the window, and (2) the events in the window need to be spilled to secondary storage. To me it seems that it still requires writing a lot of custom code. The issue might be mitigated to add flexibility on whether and where to keep the events in the window in the API.

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/topology/WindowedBoltExecutor.java:
```diff
@@ -0,0 +1,200 @@
+        Count windowLengthCount = null;
+        Duration slidingIntervalDuration = null;
+        Count slidingIntervalCount = null;
+        if (stormConf.containsKey(Config.TOPOLOGY_BOLTS_WINDOW_LENGTH_COUNT)) {
```
Given we are getting these from storm config, does this mean a user can not have 2 bolts with different windowing intervals?

arunmahadevan replied:
Its overridden by bolt level component configuration. So users can have two bolts with different intervals.

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/topology/WindowedBoltExecutor.java:
```diff
@@ -0,0 +1,200 @@
+                    TimeUnit.MILLISECONDS);
+        }
+
+        if (stormConf.containsKey(Config.TOPOLOGY_BOLTS_SLIDING_INTERVAL_COUNT)) {
```
Same as above?

arunmahadevan replied:
Its overridden by bolt level component configuration. So users can have two bolts with different intervals.

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/windowing/WindowManager.java:
```diff
@@ -0,0 +1,210 @@
+     *                  as soon as an event not satisfying the expiration policy is found.
+     * @param remaining the list of remaining events in the window after expiry.
+     */
+    private void expireEvents(boolean fullScan, List<Event<T>> remaining) {
```
why do we take a remaining list as input and mutate it? Why not just return the remaining list? same thing with expired events? Any reason for mutating instead of just returning?

arunmahadevan replied:
We dont need the remaining list always. The `expireEvents` is also called from `compactWindow` where we dont need the result and pass null. Another way would be to pass a boolean flag to control this and return the result. The `expiredEvents` is kept as a member variable since its updated during both compaction and trigger.

harshach commented:
@arunmahadevan also if the window size a bigger  lets say its 1hr how are we going to handle this. I am hoping to see a state management and a pluggable storage by default we can keep this in memory and allow user to store data in other persistent storages

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/topology/WindowedBoltExecutor.java:
```diff
@@ -0,0 +1,200 @@
+            public void onActivation(List<Tuple> tuples, List<Tuple> newTuples, List<Tuple> expiredTuples) {
+                windowedOutputCollector.setContext(tuples);
+                bolt.execute(new TupleWindowImpl(tuples, newTuples, expiredTuples));
+                newTuples.clear();
```
Again, any reason why we are relying on mutation here to clear the states, why cant the caller clear the newTuples and expiredTuples which is probably an internal state for it? 

Sorry for being anal about mutation , I am just not seeing a good reasoning to do it and it will be hard for someone else to figure out where exactly the internal state variables are changing.

arunmahadevan replied:
Yes, here the caller should clear. I think I forgot to remove this after I refactored the code a bit. Will remove this.

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/windowing/WindowManager.java:
```diff
@@ -0,0 +1,210 @@
+        Event<T> windowEvent = new EventImpl<T>(event, ts);
+        window.add(windowEvent);
+        track(windowEvent);
+        compactWindow();
```
not sure why we need this? Why not just rely on onTrigger to expire events? 

Also don't we need the locking here as well to ensure that when onTrigger is fired no new events can be added given that will change the events in window? plus i think it will give us undefined behavior as in line 187 we are using an iterator to remove an element from window list while this add method can add an element to it.

arunmahadevan replied:
Compaction is done to keep the window size in check, in case the trigger is defined such that unwanted events keep piling up in the window. Eg. window length = 100 and sliding interval is 1 min and the event rate is 1000 events/sec, without compaction, we will end storing upto 60,000 events whereas ideally we dont want to keep more than 100 events in the window. Similar cases can happen for other window configurations.

The events are stored in a `ConcurrentLinkedQueue` and onTrigger takes a snapshot of the queue and iterates over it even if elements are added meanwhile. There is locking to keep the compaction and expiry from stepping into each other. The idea was to minimize locking esp while adding events.

arunmahadevan commented:
@harshach  Right now the behavior is such that in case of failures the events will get replayed and the results will be recomputed and there will be potentially many duplicates. I assume you mean providing apis to save the partial results of computation in some persistent storage and support for retrieving the state. Along with this we also need to checkpoint the evaluated state of the window in the background and use this info to avoid re-evaluating window when events get replayed. This will help in reducing the duplicates. Since its more involved work, I will raise separate jiras to track this and will add it incrementally.

arunmahadevan commented:
@haohui Thanks for the feedback and yes your points are valid. We will incrementally add a state store for partial results and the window state and reduce window re-evaluation during replay. However, I am not sure if the tuples in window should spill over to a persistent store. I think we should rely on storm's acking mechanism to replay the events and expect that the spouts can reliable replay the events from a persistent store (e.g. kafka). This is consistent with storm's idea of not persisting in-transit tuples anywhere.

harshach commented:
@arunmahadevan my comment was regarding state storage like @haohui pointed out. Yes we can do this in follow-up jira.

arunmahadevan commented:
@harshach Storing tuples in a persistent store would mean we don't rely completely on storm's acking mechanism and build an independent logic which could complicate things. To start with we can limit the max window size to limit the memory. We can track the tuple store for windowing in a separate jira.

arunmahadevan commented:
@harshach created subtasks to track items discussed
https://issues.apache.org/jira/browse/STORM-1175 
https://issues.apache.org/jira/browse/STORM-1176

satishd commented on storm-core/src/jvm/backtype/storm/windowing/Event.java:
```diff
@@ -0,0 +1,41 @@
+     *
+     * @return
+     */
+    long getTs();
```
Can we rename getTs as getTimestamp?

arunmahadevan replied:
I used `getTs` since its short and easy to type. Yes, I think `getTimestamp` is easier to read so will rename it.

satishd commented on storm-core/src/jvm/backtype/storm/windowing/EventImpl.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Override
+    public long getTs() {
```
Can we rename getTs as getTimestamp?

revans2 commented on storm-core/src/jvm/backtype/storm/topology/base/BaseWindowedBolt.java:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * Holds a Time duration for time based windows and sliding intervals.
+     */
+    public static class Duration {
```
I don't see a lot of value in having a Count or a Duration class, Count especially.  They either wrap holding an int that is quickly unwrapped, or converting a timeUnit to millis, which too is quickly unwrapped.

revans2 replied:
I can see that they allow you to have withWindowLength method names for both, but that does not feel like it justifies adding in an extra class.  Although this is very minor, I am OK if you really want to leave it.

arunmahadevan replied:
In addition to allowing overloaded methods, it prevents accidentally passing values in wrong order if they they are just declared as ints, so I think its better to keep.

revans2 commented on storm-core/src/jvm/backtype/storm/topology/base/BaseWindowedBolt.java:
```diff
@@ -0,0 +1,180 @@
+        return withWindowLength(windowLength).withSlidingInterval(new Count(1));
+    }
+
+
```
Extra line

revans2 commented on storm-core/src/jvm/backtype/storm/windowing/EvictionPolicy.java:
```diff
@@ -0,0 +1,37 @@
+ */
+public interface EvictionPolicy<T> {
+    /**
+     * returns true if the event should be evicted from the window.
```
Could we make this a proper javadoc with an `@return` and `@param`

revans2 commented on storm-core/src/jvm/backtype/storm/windowing/TupleWindow.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * A {@link Window} that contains {@link Tuple} objects.
+ */
+public interface TupleWindow extends Window<Tuple> {
```
I'm not sure this adds a lot, this is like Count, I'm not really a huge fan of adding extra classes that don't do a lot, but this is also very minor, so if you want to keep it I am fine with it.

arunmahadevan replied:
This is just so that we can have `execute(TupleWindow window)` vs `execute(Window<Tuple> input)`

hmcl replied:
I would rather have execute(Window<Tuple> input) than execute(TupleWindow window). That is the whole purpose of generics.

revans2 commented on storm-core/src/jvm/backtype/storm/windowing/WindowManager.java:
```diff
@@ -0,0 +1,213 @@
+ */
+package backtype.storm.windowing;
+
+
```
Extra line

revans2 commented on storm-core/src/jvm/backtype/storm/windowing/WindowManager.java:
```diff
@@ -0,0 +1,213 @@
+    private List<T> expiredEvents;
+    private Set<Event<T>> prevWindowEvents;
+    private AtomicInteger eventsSinceLastExpiry;
+    private ReentrantLock lock;
```
A number of these look like they could be marked as final.

revans2 commented on storm-core/src/jvm/backtype/storm/windowing/TimeTriggerPolicy.java:
```diff
@@ -0,0 +1,78 @@
+        Runnable task = new Runnable() {
+            @Override
+            public void run() {
+                handler.onTrigger();
```
I think we need some sort of error handling here, that can pass any exceptions from the WondowedBolt back to the WindowedBoltExecutor so we can exit the worker and fail fast on unhandled exceptions.  As it stands now the exception is just eaten.

revans2 commented on storm-core/src/jvm/backtype/storm/topology/WindowedBoltExecutor.java:
```diff
@@ -0,0 +1,224 @@
+
+    @Override
+    public void execute(Tuple input) {
+        windowManager.add(input);
```
WindowManager supports both `add(Tuple)` and `add(Tuple, timestamp)` the second API is tested, but it is never called.  Is this just for testing?  If so it is a real shame, because I know of many cases where a we want to window on the data at event creation time, not at the time it arrived at the bolt.  Additionally there are a few places in the code that refer to dealing with late data, but I don't see how anything can be "late" or out of order without having access to that second method.  If it is intended to just be a test API please file a follow on JIRA so that we can expose it to end users in some way and test with out-of-order data.  If it was not intended to be a test method, please provide a way to set it, preferably through an API in IWindowedBolt.

arunmahadevan replied:
I added it with the idea of supporting late events, but realized that it will need more work and thought it can be added once the basic windowing is in place. I will file a follow on jira and work on it.

revans2 commented:
I when through the code in more detail and I had a few minor nits, but also a few issues that I am concerned about.

Primarily error handling when using a separate thread for timed expiry. But fundamentally I am concerned that the windowing functionality assumes that arrival time to the bolt is the only time that people will want to do windowed intervals over, which is just not true.

This potentially opens up a whole can of worms the more I think about it, and I am not sure that this is the best way to do windowing at scale if we want to be able to support late data, out of order arrival, and event creation time as a part of the window.

Additionally I would like to see more documentation about how to use this functionality, not just a simple example.

revans2 commented:
As a note on the late data and out of order processing please take a look at the Google Could Data Flow API.  They have been dealing with these types of use cases for a long time, and have developed an API, although a bit ugly, that really captures the essence of this type of a problem.

arunmahadevan commented:
@revans2 I have added error handling to the scheduled executor used for time based expiry and addressed most of your other comments. I feel late data and out of order processing can be done incrementally once basic windowing support is added. I will track it in a separate jira - https://issues.apache.org/jira/browse/STORM-1187

revans2 commented:
@arunmahadevan I am OK with adding in support for other feature incrementally, but I am a bit concerned that the API will have to change fundamentally to be able to support these types of processing, especially at scale, but we can cross that bridge when we come to it, and if we need a different API we can add it and deprecate the old.

I am +1, but I would like to hear from @harshach before merging this in.

harshach commented:
@arunmahadevan @Parth-Brahmbhatt @revans2 still going through the code, give me a day i'll put my comments in. Thanks.

revans2 commented:
@harshach sure I am not in a rush.

revans2 commented:
@harshach any update on your review?

harshach commented:
+1. Sorry for the delay.

ptgoetz commented on storm-core/src/jvm/backtype/storm/topology/TopologyBuilder.java:
```diff
@@ -177,6 +179,21 @@ public BoltDeclarer setBolt(String id, IBasicBolt bolt, Number parallelism_hint)
+     * @return use the returned object to declare the inputs to this component
+     * @throws IllegalArgumentException if {@code parallelism_hint} is not positive
+     */
+    public BoltDeclarer setBolt(String id, IWindowedBolt bolt, Number parallelism_hint) throws IllegalArgumentException {
```
This API addition will require changes in Flux in order for it to support windowing.

We should file a JIRA to track that.

arunmahadevan replied:
Thanks. I 've added a tracking JIRA https://issues.apache.org/jira/browse/STORM-1207

harshach commented:
Thanks @arunmahadevan merged into trunk.

hmcl commented on storm-core/src/jvm/backtype/storm/windowing/EventImpl.java:
```diff
@@ -0,0 +1,38 @@
+ */
+package backtype.storm.windowing;
+
+class EventImpl<T> implements Event<T> {
```
Is there a need to have the class EventImpl and the interface Event? Is it reasonable to think that there will ever be another implementation of Event. If not, which in my opinion is the case, we should have only the class Event which provides the implementation. No need for the interface. In the article pasted bellow it says that a good  for when there should be no need for an interface and an implementing class is when one cannot come up with a better name for the class rather than adding the suffix Impl.

http://stackoverflow.com/questions/2814805/java-interfaces-implementation-naming-convention

arunmahadevan replied:
Agree with your comments for value types and final classes. When I wrote, I was not sure whether we would need another implementation and it turns out that we need another implementation (watermark events) for tracking tuple timestamps.
If unsure I feel it might be better to err by having interfaces. It makes it easier to write mock implementation for testing and makes the apis more readable without cluttering with the implementation details. (see below link for more details).

http://programmers.stackexchange.com/questions/159813/do-i-need-to-use-an-interface-when-only-one-class-will-ever-implement-it/159909#159909

hmcl commented on storm-core/src/jvm/backtype/storm/windowing/TupleWindowImpl.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * Holds the expired, new and current tuples in a window.
+ */
+public class TupleWindowImpl implements TupleWindow {
```
Is there a need to have the class TupleWindowImpl and the interface TupleWindow? For the reasons in the comment above.
